In [6]:
import pandas as pd 
import csv 
import geopandas as gpd
import os
pd.set_option('display.max_columns', None) # to display max columns of the dataframe

- All datasets used in this script is taken from the FDIC API
- This link contains the API Documentation including feature descriptions: https://banks.data.fdic.gov/docs/#/

### Read locations csv files and create deep copy
- The locations csv files contains all financial institutions in the US including US owned branches outside of the US

In [1932]:
# locations data 
locations_csv_path = '../data_collection/locations_data.csv'
locations_df_orig = pd.read_csv(locations_csv_path)

# institutions data 
institutions_csv_path = '../data_collection/institutions_data.csv'
institutions_df_orig = pd.read_csv(institutions_csv_path)

# summary of deposits data
sod_csv_path = '../data_collection/sod_data.csv'
sod_df_orig = pd.read_csv(sod_csv_path)

C:\Users\jocel\AppData\Local\Temp\ipykernel_51800\242835875.py:3: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  locations_df_orig = pd.read_csv(locations_csv_path)


In [1933]:
# create a deep copy so original dataframe does not get altered
locations_df = locations_df_orig.copy(deep = True)
institutions_df = institutions_df_orig.copy(deep = True)
sod_df = sod_df_orig.copy(deep = True)

### Locations data

In [1935]:
# locations dataframe 
# if an instance's mainoff = 0, NAME is the Main Office Name and OFFNUM and OFFNUM is the bank branch number and name
# mainoff = 1 is a MAIN OFFICE therefore OFFNAME == NAME, and OFFNUM is 0 
locations_df.head(20)

ADDRESS BKCLASS                                 CBSA  \
0          102 Rogers St       N       Boston-Cambridge-Newton, MA-NH   
1       401 E Jackson St      NM  Tampa-St. Petersburg-Clearwater, FL   
2           117 N 3rd St      NM                         Mitchell, SD   
3       803 Wollard Blvd      NM                   Kansas City, MO-KS   
4      706 Washington St      NM                           Ottawa, IL   
5            410 14th St      NM                                  NaN   
6            10 W Elm St       N                                  NaN   
7       6340 Merchant St      NM                                  NaN   
8   301 West Main Street      NM                              Ada, OK   
9         122 W State St       N                                  NaN   
10           459 Main St      SM                         Bay City, TX   
11         248 W Main St      NM                         Florence, SC   
12         113 S Main St      NM                                  NaN   
13           308 Main St      NM                       Mason City, IA   
14        101 N Duval St      NM                                  NaN   
15        1512 Ogden Ave       N   Chicago-Naperville-Elgin, IL-IN-WI   
16      12461 Augusta Rd      NM                                  NaN   
17           605 Main St      NM                    Burlington, IA-IL   
18          100 E 2nd St      SM                          Norfolk, NE   
19          205 Main Ave      NM                                  NaN   

                           CBSA_DIV  CBSA_DIV_FLG  CBSA_DIV_NO  CBSA_METRO  \
0   Cambridge-Newton-Framingham, MA             1      15764.0       14460   
1                               NaN             0          NaN       45300   
2                               NaN             0          NaN           0   
3                               NaN             0          NaN       28140   
4                               NaN             0          NaN           0   
5                               NaN             0          0.0           0   
6                               NaN             0          0.0           0   
7                               NaN             0          0.0           0   
8                               NaN             0          NaN           0   
9                               NaN             0          0.0           0   
10                              NaN             0          NaN           0   
11                              NaN             0          NaN       22500   
12                              NaN             0          0.0           0   
13                              NaN             0          NaN           0   
14                              NaN             0          0.0           0   
15  Chicago-Naperville-Evanston, IL             1      16984.0       16980   
16                              NaN             0          0.0           0   
17                              NaN             0          NaN           0   
18                              NaN             0          NaN           0   
19                              NaN             0          0.0           0   

    CBSA_METRO_FLG                      CBSA_METRO_NAME  CBSA_MICRO_FLG  \
0                1       Boston-Cambridge-Newton, MA-NH               0   
1                1  Tampa-St. Petersburg-Clearwater, FL               0   
2                0                                  NaN               1   
3                1                   Kansas City, MO-KS               0   
4                0                                  NaN               1   
5                0                                  NaN               0   
6                0                                  NaN               0   
7                0                                  NaN               0   
8                0                                  NaN               1   
9                0                                  NaN               0   
10               0 

In [1936]:
locations_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78891 entries, 0 to 78890
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ADDRESS          78891 non-null  object 
 1   BKCLASS          78891 non-null  object 
 2   CBSA             69156 non-null  object 
 3   CBSA_DIV         17802 non-null  object 
 4   CBSA_DIV_FLG     78891 non-null  int64  
 5   CBSA_DIV_NO      27537 non-null  float64
 6   CBSA_METRO       78891 non-null  int64  
 7   CBSA_METRO_FLG   78891 non-null  int64  
 8   CBSA_METRO_NAME  60370 non-null  object 
 9   CBSA_MICRO_FLG   78891 non-null  int64  
 10  CBSA_NO          78891 non-null  int64  
 11  CITY             78891 non-null  object 
 12  COUNTY           78091 non-null  object 
 13  ESTYMD           78891 non-null  object 
 14  ID               78891 non-null  int64  
 15  MAINOFF          78891 non-null  int64  
 16  NAME             78891 non-null  object 
 17  OFFNAME     

In [1937]:
# UNINUM is a unique identifier for the locations dataset 
unique_num_l = locations_df['UNINUM'].nunique()
print(f'There are {unique_num_l} unique identifiers in the locations dataset')

There are 78891 unique identifiers in the locations dataset


In [1938]:
# dropping all CBSA columns 
locations_df.drop(columns = ['CBSA', 'CBSA_DIV', 'CBSA_DIV_FLG', 'CBSA_DIV_NO', 'CBSA_METRO', 'CBSA_METRO_FLG', 'CBSA_METRO_NAME', 'CBSA_MICRO_FLG', 'CBSA_NO'], inplace = True)

In [1939]:
locations_df.head()

ADDRESS BKCLASS        CITY        COUNTY      ESTYMD     ID  \
0      102 Rogers St       N  Gloucester         Essex  01/01/1891     10   
1   401 E Jackson St      NM       Tampa  Hillsborough  07/01/1914    100   
2       117 N 3rd St      NM       Emery        Hanson  01/01/1920  10000   
3   803 Wollard Blvd      NM    Richmond           Ray  07/01/1940  10002   
4  706 Washington St      NM     Mendota       Lasalle  07/06/1940  10003   

   MAINOFF                     NAME                         OFFNAME  OFFNUM  \
0        0     Santander Bank, N.A.               GLOUCESTER BRANCH     679   
1        0              Truist Bank  TRUIST FINANCIAL CENTER BRANCH    3378   
2        1  The Security State Bank         The Security State Bank       0   
3        1         Flat Branch Bank                Flat Branch Bank       0   
4        1         First State Bank                First State Bank       0   

   SERVTYPE STALP  STCNTY         STNAME  UNINUM    ZIP  
0        11    MA   25009  Massachusetts      10  01930  
1        11    FL   12057        Florida     100  33602  
2        11    SD   46061   South Dakota   10000  57332  
3        11    MO   29177       Missouri   10002  64085  
4        11    IL   17099       Illinois   10003  61342

### Geocoding addresses from locations csv file
- The purpose of using the geocoding services web API of the US Census Bureau is to geocode the addresses available in the locations data scraped from the FDIC API
- Priority to extract Coordinates, State, County and Tract Code for each address

##### Explanation of code below 
- Scraped locations data from the FDIC api that contains locations of all main banks and branches (inside and outside the US), this data does not contain a census tract code and coordinates
- In order to use the Us Census Bureau's Geocoding Services Web API (https://geocoding.geo.census.gov/geocoder/Geocoding_Services_API.html), the csv file to be geocoded has to follow a certain structure: Unique ID, Street address, City, State, ZIP so I created address_for_geo csv file that contains all these columns to prepare for geocoding and split this csv file into 9 different csv files due to the api's batch geocoding limit restrictions of 10000
- After the preparing the csv files, I used the requests package and sent a request to the geocoding api, this portion was done in the data collection notebook
- After geocoding all the addresses and saving all the output files, I concatenated all the output csv files into one file: addresses_geocoded_df.csv
- The concatenation process was followed by some basic practices when data cleaning, checking null values, dropping nulls, dropping duplicates and changing the data type of State, County, Tract and Block code to str
- The csv file used for geocoding initially had 78891 rows but after geocoding, the csv file is left with 68240 rows (some of the addresses were located outside of the US, these are US owned bank branches outside of the US. The geocoding services api only geocodes locations inside the US). Due to this, the data was filtered and now only includes bank locations in the United States.

In [1941]:
# use geocoding web services api by the us census bureau 
# create a function for batch geocoding of addresses in the locations data file 
# batch geocoding - put addresses in a csv file - needs to be in this format: Unique ID, Street address, City, State, ZIP

address_for_geo = locations_df[['UNINUM', 'ADDRESS', 'CITY', 'STNAME', 'ZIP']]
print(address_for_geo)
# address_for_geo.to_csv('address_for_geo.csv', index = False)

       UNINUM              ADDRESS         CITY         STNAME    ZIP
0          10        102 Rogers St   Gloucester  Massachusetts  01930
1         100     401 E Jackson St        Tampa        Florida  33602
2       10000         117 N 3rd St        Emery   South Dakota  57332
3       10002     803 Wollard Blvd     Richmond       Missouri  64085
4       10003    706 Washington St      Mendota       Illinois  61342
...       ...                  ...          ...            ...    ...
78886    9993         607 6th St S  Springville           Iowa  52336
78887    9994  1251 Us Highway 70a       Wilson       Oklahoma  73463
78888    9995        325 Brown Ave         Mott   North Dakota  58646
78889    9996    321 W Francis Ave        Pampa          Texas  79065
78890    9997    139 E Broadway St  Philipsburg        Montana  59858

[78891 rows x 5 columns]


In [1942]:
# since the geocoder only limits 10000 records per batch file, we will split the csv into batches

# address = 'address_geo.csv'
# chunk_size = 9000

# for i, chunk in enumerate(range(0, len(address_for_geo), chunk_size)):
#     output_file = f"addresses_batch_{i+1}.csv"
#     address_for_geo.iloc[chunk : chunk + chunk_size].to_csv(output_file, index=False)
#     print(f"Saved {output_file}")

In [1943]:
# read all geocoded addresses files and put it in a combined list
# geocoded addresses script is in data_collection

# geocoded_files = []

# column_names = [
#     "UNINUM", "Input Address", "Match Type", "Match Precision",
#     "Standardized Address", "Coordinates", "Census Geoid", 
#     "Unknown1", "State Code", "County Code", "Tract Code", "Block Code"
# ]
# # the address_geo csv files were split into 9 due to the geocoder's batch limit restrictions
# for i in range(1, 10):
#     geocoded_csv = f'addresses_geocoded_{i}.csv'

#     try: 
#         geo_df = pd.read_csv(geocoded_csv, header = None, names = column_names, dtype = str, keep_default_na = False)

#         # filter addresses that got NO_MATCH values due to the address being outside of the US
#         geo_df = geo_df[geo_df['Match Type'] != 'No_Match']
        
#         geo_df['State Code'] = geo_df['State Code'].apply(lambda x: str(x).zfill(2)) # state code must be 2 digits 
#         geo_df['County Code'] = geo_df['County Code'].apply(lambda x: str(x).zfill(3)) # county code must be 3 digits
#         geo_df['Tract Code'] = geo_df['Tract Code'].apply(lambda x: str(int(x)).zfill(6) if x.strip() != '' else '000000') # tract code must be 6 digits
#         geo_df['Block Code'] = geo_df['Block Code'].apply(lambda x: str(x).zfill(4)) # block code must be 4 digits
        
#         geocoded_files.append(geo_df)
#     except FileNotFoundError as e: 
#         print('Warning: File not Found')

# # concatenate all csv files from geocoded_files list 
# if geocoded_files:
#     geocoded_combined = pd.concat(geocoded_files, ignore_index = True)
#     geocoded_combined.to_csv('addresses_geocoded_final.csv', index = False, header = True)  
#     print("Successfully saved 'addresses_geocoded_final.csv'.")
# else:
#     print('No files found to concatenate.')

In [1944]:
# concatenate the addresses_geocoded_final_csv to the locations_df above 
# pandas keep dropping the 0's in Tract Code columns - needs to be a 6 digit code
addresses_geocoded_df = pd.read_csv("addresses_geocoded_final.csv")
addresses_geocoded_df

UNINUM                                      Input Address Match Type  \
0      184855  2200 Cottman Ave, Philadelphia, Pennsylvania, ...      Match   
1      184854  4356 Frankford Ave, Philadelphia, Pennsylvania...      Match   
2      183527  1290 Avenue Of The Americas, New York, New Yor...      Match   
3       17285  978 W San Marcos Blvd, San Marcos, California,...      Match   
4       17286           6501 Hillcrest Ave, Dallas, Texas, 75205      Match   
...       ...                                                ...        ...   
69059  657640  100 Centre Street, Brookline, Massachusetts, 2466      Match   
69060  657641         6400 Rockside Road, Cleveland, Ohio, 44131      Match   
69061  657645      275 Dock Drive, Lansdale, Pennsylvania, 19446      Match   
69062  657642     1240 N Bridge Street, Chillicothe, Ohio, 45601      Match   
69063  657643    5510 S. Interstate Hwy 35, Austin, Texas, 78745      Match   

      Match Precision                           Standardized Address  \
0               Exact      2200 COTTMAN AVE, PHILADELPHIA, PA, 19149   
1               Exact    4356 FRANKFORD AVE, PHILADELPHIA, PA, 19124   
2           Non_Exact  1290 AVE OF THE AMERICAS, NEW YORK, NY, 10104   
3               Exact   978 W SAN MARCOS BLVD, SAN MARCOS, CA, 92078   
4               Exact          6501 HILLCREST AVE, DALLAS, TX, 75205   
...               ...                                            ...   
69059       Non_Exact            100 CENTRE ST, BROOKLINE, MA, 02446   
69060           Exact         6400 ROCKSIDE RD, CLEVELAND, OH, 44131   
69061           Exact               275 DOCK DR, LANSDALE, PA, 19446   
69062           Exact       1240 N BRIDGE ST, CHILLICOTHE, OH, 45601   
69063       Non_Exact         5510 S I- 35 SVC RD, AUSTIN, TX, 78745   

                             Coordinates  Census Geoid Unknown1  State Code  \
0       -75.062366464517,40.049065909735   131416528.0        R          42   
1       -75.088234485579,40.011821629265   131427808.0        L          42   
2        -73.97965465923,40.760835189154    59657151.0        R          36   
3      -117.185996915856,33.135366914995   195295827.0        R           6   
4       -96.787090572635,32.845309044897   102907323.0        L          48   
...                                  ...           ...      ...         ...   
69059   -71.126158060077,42.343556166673    87244381.0        L          25   
69060   -81.647118261688,41.395264086175   638893764.0        R          39   
69061   -75.336524536217,40.257154935469   605639075.0        L          42   
69062   -82.976674170295,39.351349664134    37495157.0        L          39   
69063   -97.763284269579,30.200202510377    63969786.0        R          48   

       County Code  Tract Code  Block Code  
0              101       31401        3005  
1              101       29300        2006  
2               61       10400        1004  
3               73       20029        4026  
4              113       19400        2037  
...            ...         ...         ...  
69059           21      400401        1001  
69060           35      156101        1047  
69061           91      206904        3018  
69062          141      956300        2034  
69063          453        2443        1009  

[69064 rows x 12 columns]

In [1945]:
# number of unique UNINUMS that had a match in the geocoding process
addresses_geocoded_df['UNINUM'].nunique()

69064

In [1946]:
# duplicate rows in the addresses_geocoded_df
# no duplicate rows
addresses_geocoded_df.duplicated().sum()

0

In [1947]:
addresses_geocoded_df.shape

(69064, 12)

In [1948]:
# check null values for each columns 
addresses_geocoded_df.isna().sum()

UNINUM                    0
Input Address             0
Match Type                0
Match Precision         570
Standardized Address    570
Coordinates             570
Census Geoid            824
Unknown1                824
State Code                0
County Code               0
Tract Code                0
Block Code                0
dtype: int64

In [1949]:
# drop all rows with NaN value 
addresses_geocoded_df.dropna(inplace = True)

In [1950]:
addresses_geocoded_df

UNINUM                                      Input Address Match Type  \
0      184855  2200 Cottman Ave, Philadelphia, Pennsylvania, ...      Match   
1      184854  4356 Frankford Ave, Philadelphia, Pennsylvania...      Match   
2      183527  1290 Avenue Of The Americas, New York, New Yor...      Match   
3       17285  978 W San Marcos Blvd, San Marcos, California,...      Match   
4       17286           6501 Hillcrest Ave, Dallas, Texas, 75205      Match   
...       ...                                                ...        ...   
69059  657640  100 Centre Street, Brookline, Massachusetts, 2466      Match   
69060  657641         6400 Rockside Road, Cleveland, Ohio, 44131      Match   
69061  657645      275 Dock Drive, Lansdale, Pennsylvania, 19446      Match   
69062  657642     1240 N Bridge Street, Chillicothe, Ohio, 45601      Match   
69063  657643    5510 S. Interstate Hwy 35, Austin, Texas, 78745      Match   

      Match Precision                           Standardized Address  \
0               Exact      2200 COTTMAN AVE, PHILADELPHIA, PA, 19149   
1               Exact    4356 FRANKFORD AVE, PHILADELPHIA, PA, 19124   
2           Non_Exact  1290 AVE OF THE AMERICAS, NEW YORK, NY, 10104   
3               Exact   978 W SAN MARCOS BLVD, SAN MARCOS, CA, 92078   
4               Exact          6501 HILLCREST AVE, DALLAS, TX, 75205   
...               ...                                            ...   
69059       Non_Exact            100 CENTRE ST, BROOKLINE, MA, 02446   
69060           Exact         6400 ROCKSIDE RD, CLEVELAND, OH, 44131   
69061           Exact               275 DOCK DR, LANSDALE, PA, 19446   
69062           Exact       1240 N BRIDGE ST, CHILLICOTHE, OH, 45601   
69063       Non_Exact         5510 S I- 35 SVC RD, AUSTIN, TX, 78745   

                             Coordinates  Census Geoid Unknown1  State Code  \
0       -75.062366464517,40.049065909735   131416528.0        R          42   
1       -75.088234485579,40.011821629265   131427808.0        L          42   
2        -73.97965465923,40.760835189154    59657151.0        R          36   
3      -117.185996915856,33.135366914995   195295827.0        R           6   
4       -96.787090572635,32.845309044897   102907323.0        L          48   
...                                  ...           ...      ...         ...   
69059   -71.126158060077,42.343556166673    87244381.0        L          25   
69060   -81.647118261688,41.395264086175   638893764.0        R          39   
69061   -75.336524536217,40.257154935469   605639075.0        L          42   
69062   -82.976674170295,39.351349664134    37495157.0        L          39   
69063   -97.763284269579,30.200202510377    63969786.0        R          48   

       County Code  Tract Code  Block Code  
0              101       31401        3005  
1              101       29300        2006  
2               61       10400        1004  
3               73       20029        4026  
4              113       19400        2037  
...            ...         ...         ...  
69059           21      400401        1001  
69060           35      156101        1047  
69061           91      206904        3018  
69062          141      956300        2034  
69063          453        2443        1009  

[68240 rows x 12 columns]

In [1951]:
# no remaining NaN values 
addresses_geocoded_df.isna().sum()

UNINUM                  0
Input Address           0
Match Type              0
Match Precision         0
Standardized Address    0
Coordinates             0
Census Geoid            0
Unknown1                0
State Code              0
County Code             0
Tract Code              0
Block Code              0
dtype: int64

In [1952]:
# Reset the index of the DataFrame
addresses_geocoded_df.reset_index(drop=True, inplace=True)

In [1953]:
addresses_geocoded_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68240 entries, 0 to 68239
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   UNINUM                68240 non-null  int64  
 1   Input Address         68240 non-null  object 
 2   Match Type            68240 non-null  object 
 3   Match Precision       68240 non-null  object 
 4   Standardized Address  68240 non-null  object 
 5   Coordinates           68240 non-null  object 
 6   Census Geoid          68240 non-null  float64
 7   Unknown1              68240 non-null  object 
 8   State Code            68240 non-null  int64  
 9   County Code           68240 non-null  int64  
 10  Tract Code            68240 non-null  int64  
 11  Block Code            68240 non-null  int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 6.2+ MB


In [1954]:
# change data type of state, county, tract and block code to str
addresses_geocoded_df['State Code'] = addresses_geocoded_df['State Code'].astype(str)
addresses_geocoded_df['County Code'] = addresses_geocoded_df['County Code'].astype(str)
addresses_geocoded_df['Tract Code'] = addresses_geocoded_df['Tract Code'].astype(str)
addresses_geocoded_df['Block Code'] = addresses_geocoded_df['Block Code'].astype(str)

# filling the tract code columns with zeroes until it's 6 digits long 
addresses_geocoded_df['Tract Code'] = addresses_geocoded_df['Tract Code'].apply(lambda x: x.zfill(6))
addresses_geocoded_df['State Code'] = addresses_geocoded_df['State Code'].apply(lambda x: x.zfill(2))
addresses_geocoded_df['County Code'] = addresses_geocoded_df['County Code'].apply(lambda x: x.zfill(3))

Prioritized keeping rows with coordinates, state, county, tract and block code.

### Feature Engineering geocoded addresses from locations dataset
- Splitting Coordinates column and extracting the Longitude and Latitude
- Creating an 11 digit FIPS code (combined state, country and tract code), this FIPS code will be useful in defining whether a particular census tract is an existing bank desert or a potential bank desert

In [1957]:
# feature engineer the COORDINATES columns to create Longitude and Latitude columns 

# Longitude column
addresses_geocoded_df['Longitude'] = addresses_geocoded_df['Coordinates'].apply(lambda x: x.split(',')[0])

# Latitude column
addresses_geocoded_df['Latitude'] = addresses_geocoded_df['Coordinates'].apply(lambda x: x.split(',')[1])

In [1958]:
# feature engineer a FULL 11 DIGIT FIPS CODE (state, county, tract code)
# applying lambda to multiple columns at once without specifying a single columns and using axis = 1

addresses_geocoded_df['FIPS 11'] = addresses_geocoded_df.apply(lambda row: row['State Code'] + row['County Code'] + row['Tract Code'], axis = 1)

In [1959]:
addresses_geocoded_df

UNINUM                                      Input Address Match Type  \
0      184855  2200 Cottman Ave, Philadelphia, Pennsylvania, ...      Match   
1      184854  4356 Frankford Ave, Philadelphia, Pennsylvania...      Match   
2      183527  1290 Avenue Of The Americas, New York, New Yor...      Match   
3       17285  978 W San Marcos Blvd, San Marcos, California,...      Match   
4       17286           6501 Hillcrest Ave, Dallas, Texas, 75205      Match   
...       ...                                                ...        ...   
68235  657640  100 Centre Street, Brookline, Massachusetts, 2466      Match   
68236  657641         6400 Rockside Road, Cleveland, Ohio, 44131      Match   
68237  657645      275 Dock Drive, Lansdale, Pennsylvania, 19446      Match   
68238  657642     1240 N Bridge Street, Chillicothe, Ohio, 45601      Match   
68239  657643    5510 S. Interstate Hwy 35, Austin, Texas, 78745      Match   

      Match Precision                           Standardized Address  \
0               Exact      2200 COTTMAN AVE, PHILADELPHIA, PA, 19149   
1               Exact    4356 FRANKFORD AVE, PHILADELPHIA, PA, 19124   
2           Non_Exact  1290 AVE OF THE AMERICAS, NEW YORK, NY, 10104   
3               Exact   978 W SAN MARCOS BLVD, SAN MARCOS, CA, 92078   
4               Exact          6501 HILLCREST AVE, DALLAS, TX, 75205   
...               ...                                            ...   
68235       Non_Exact            100 CENTRE ST, BROOKLINE, MA, 02446   
68236           Exact         6400 ROCKSIDE RD, CLEVELAND, OH, 44131   
68237           Exact               275 DOCK DR, LANSDALE, PA, 19446   
68238           Exact       1240 N BRIDGE ST, CHILLICOTHE, OH, 45601   
68239       Non_Exact         5510 S I- 35 SVC RD, AUSTIN, TX, 78745   

                             Coordinates  Census Geoid Unknown1 State Code  \
0       -75.062366464517,40.049065909735   131416528.0        R         42   
1       -75.088234485579,40.011821629265   131427808.0        L         42   
2        -73.97965465923,40.760835189154    59657151.0        R         36   
3      -117.185996915856,33.135366914995   195295827.0        R         06   
4       -96.787090572635,32.845309044897   102907323.0        L         48   
...                                  ...           ...      ...        ...   
68235   -71.126158060077,42.343556166673    87244381.0        L         25   
68236   -81.647118261688,41.395264086175   638893764.0        R         39   
68237   -75.336524536217,40.257154935469   605639075.0        L         42   
68238   -82.976674170295,39.351349664134    37495157.0        L         39   
68239   -97.763284269579,30.200202510377    63969786.0        R         48   

      County Code Tract Code Block Code          Longitude         Latitude  \
0             101     031401       3005   -75.062366464517  40.049065909735   
1             101     029300       2006   -75.088234485579  40.011821629265   
2             061     010400       1004    -73.97965465923  40.760835189154   
3             073     020029       4026  -117.185996915856  33.135366914995   
4             113     019400       2037   -96.787090572635  32.845309044897   
...           ...        ...        ...                ...              ...   
68235         021     400401       1001   -71.126158060077  42.343556166673   
68236         035     156101       1047   -81.647118261688  41.395264086175   
68237         091     206904       3018   -75.336524536217  40.257154935469   
68238         141     956300       2034   -82.976674170295  39.351349664134   
68239         453     002443       1009   -97.763284269579  30.200202510377   

           FIPS 11  
0      42101031401  
1      42101029300  
2      36061010400  
3      06073020029  
4      48113019400  
...            ...  
68235  25021400401  
68236  39035156101  
68237  42091206904  
68238  39141956300  
68239  48453002443  

[68240 rows x 15 columns]

In [1960]:
# check the columns in locations_data and addresses_geocoded_df to see if we need to rename some columns 

location_df_cols = locations_df.columns
geocoded_df_cols = addresses_geocoded_df.columns 
print(f'Locations data columns: {location_df_cols}')
print(f'Geocoded data columns: {geocoded_df_cols}')

Locations data columns: Index(['ADDRESS', 'BKCLASS', 'CITY', 'COUNTY', 'ESTYMD', 'ID', 'MAINOFF',
       'NAME', 'OFFNAME', 'OFFNUM', 'SERVTYPE', 'STALP', 'STCNTY', 'STNAME',
       'UNINUM', 'ZIP'],
      dtype='object')
Geocoded data columns: Index(['UNINUM', 'Input Address', 'Match Type', 'Match Precision',
       'Standardized Address', 'Coordinates', 'Census Geoid', 'Unknown1',
       'State Code', 'County Code', 'Tract Code', 'Block Code', 'Longitude',
       'Latitude', 'FIPS 11'],
      dtype='object')


### Merging locations data with geocoded address

In [1962]:
# merging locations data and geocoded addresses on UNINUM 

loc_geo_df = pd.merge(locations_df, addresses_geocoded_df, how = 'inner', on = 'UNINUM')

In [1963]:
# saving this dataframe to a csv file 

# loc_geo_df.to_csv('loc_geo_final.csv', index = False)

In [1964]:
# dropping repetitive and unnecessary columns

loc_geo_df.drop(columns = ['ID', 'SERVTYPE', 'Match Type', 'Match Precision', 'Standardized Address', 'Coordinates', 'Unknown1'], inplace = True)

In [1965]:
loc_geo_df

ADDRESS BKCLASS         CITY        COUNTY      ESTYMD  \
0            102 Rogers St       N   Gloucester         Essex  01/01/1891   
1         401 E Jackson St      NM        Tampa  Hillsborough  07/01/1914   
2         803 Wollard Blvd      NM     Richmond           Ray  07/01/1940   
3        706 Washington St      NM      Mendota       Lasalle  07/06/1940   
4              410 14th St      NM   Burlington    Kit Carson  01/10/2023   
...                    ...     ...          ...           ...         ...   
68235         607 6th St S      NM  Springville          Linn  06/02/2023   
68236  1251 Us Highway 70a      SM       Wilson        Carter  05/16/1940   
68237        325 Brown Ave      NM         Mott     Hettinger  06/12/1940   
68238    321 W Francis Ave      NM        Pampa          Gray  05/27/1940   
68239    139 E Broadway St      NM  Philipsburg       Granite  06/02/2023   

       MAINOFF                               NAME  \
0            0               Santander Bank, N.A.   
1            0                        Truist Bank   
2            1                   Flat Branch Bank   
3            1                   First State Bank   
4            0  Farmers & Merchants Bank of Colby   
...        ...                                ...   
68235        0                Citizens State Bank   
68236        0               American Nation Bank   
68237        1            Commercial Bank of Mott   
68238        0                FirstBank Southwest   
68239        0             Citizens Alliance Bank   

                              OFFNAME  OFFNUM STALP  STCNTY         STNAME  \
0                   GLOUCESTER BRANCH     679    MA   25009  Massachusetts   
1      TRUIST FINANCIAL CENTER BRANCH    3378    FL   12057        Florida   
2                    Flat Branch Bank       0    MO   29177       Missouri   
3                    First State Bank       0    IL   17099       Illinois   
4                   BURLINGTON BRANCH       6    CO    8063       Colorado   
...                               ...     ...   ...     ...            ...   
68235              SPRINGVILLE BRANCH       8    IA   19113           Iowa   
68236                   WILSON BRANCH       2    OK   40019       Oklahoma   
68237         Commercial Bank of Mott       0    ND   38041   North Dakota   
68238                    PAMPA BRANCH       4    TX   48179          Texas   
68239              PHILIPSBURG BRANCH      20    MT   30039        Montana   

       UNINUM    ZIP                                    Input Address  \
0          10  01930  102 Rogers St, Gloucester, Massachusetts, 01930   
1         100  33602          401 E Jackson St, Tampa, Florida, 33602   
2       10002  64085      803 Wollard Blvd, Richmond, Missouri, 64085   
3       10003  61342      706 Washington St, Mendota, Illinois, 61342   
4       10007  80807         410 14th St, Burlington, Colorado, 80807   
...       ...    ...                                              ...   
68235    9993  52336           607 6th St S, Springville, Iowa, 52336   
68236    9994  73463     1251 Us Highway 70a, Wilson, Oklahoma, 73463   
68237    9995  58646         325 Brown Ave, Mott, North Dakota, 58646   
68238    9996  79065           321 W Francis Ave, Pampa, Texas, 79065   
68239    9997  59858   139 E Broadway St, Philipsburg, Montana, 59858   

       Census Geoid State Code County Code Tract Code Block Code  \
0        86668295.0         25         009     221500       2005   
1       104530591.0         12         057     005101       2079   
2        20986846.0         29         177     080202       2002   
3         8582457.0         17         099     961900       4022   
4       132236547.0         08         063     962100       2065   
...             ...        ...         ...        ...        ...   
68235    17014761.0         19         113     010400       1045   
68236    14825623.0         40         019     892400       3007   
68237   197188635.0         38 

In [1966]:
loc_geo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68240 entries, 0 to 68239
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ADDRESS        68240 non-null  object 
 1   BKCLASS        68240 non-null  object 
 2   CITY           68240 non-null  object 
 3   COUNTY         68234 non-null  object 
 4   ESTYMD         68240 non-null  object 
 5   MAINOFF        68240 non-null  int64  
 6   NAME           68240 non-null  object 
 7   OFFNAME        68240 non-null  object 
 8   OFFNUM         68240 non-null  int64  
 9   STALP          68234 non-null  object 
 10  STCNTY         68240 non-null  int64  
 11  STNAME         68234 non-null  object 
 12  UNINUM         68240 non-null  int64  
 13  ZIP            68240 non-null  object 
 14  Input Address  68240 non-null  object 
 15  Census Geoid   68240 non-null  float64
 16  State Code     68240 non-null  object 
 17  County Code    68240 non-null  object 
 18  Tract 

In [1967]:
# 14289 unique tract codes, which means that some census tracts contain more than 1 bank 
loc_geo_df['Tract Code'].nunique()

14289

### Rural Urban Commuting Area Code 
- The rural-urban commuting area (RUCA) codes classify U.S. census tracts using measures of population density, urbanization, and daily commuting
- The ruca code from the USDA can be found here: https://www.ers.usda.gov/data-products/rural-urban-commuting-area-codes#:~:text=The%20rural%2Durban%20commuting%20area,%2C%20urbanization%2C%20and%20daily%20commuting. 
- The Primary RUCA code classification descriptions can be found here: https://www.ers.usda.gov/data-products/rural-urban-commuting-area-codes/documentation
- This data will be used to feature engineer the rural/urban classification (Community Type) depending on the census tract code, this feature will help identify whether a particular census tract is a potential bank desert, an existing bank desert or not a bank desert. 

In [1969]:
# read ruca codes excel file and transform to a csv file 
# ruca = pd.read_excel('ruca2010revised.xlsx')

# writing ruca excel file to csv file 
# ruca.to_csv('ruca_codes.csv', index = False)

In [1970]:
# read ruca csv data and create a deep copy 
ruca_csv = pd.read_csv('ruca_codes.csv', low_memory = False, skiprows = 1)
ruca_df = ruca_csv.copy(deep = True)

In [1971]:
ruca_df.head()

State-County FIPS Code Select State   Select County  \
0                    1001           AL  Autauga County   
1                    1001           AL  Autauga County   
2                    1001           AL  Autauga County   
3                    1001           AL  Autauga County   
4                    1001           AL  Autauga County   

   State-County-Tract FIPS Code (lookup by address at http://www.ffiec.gov/Geocode/)  \
0                                         1001020100                                   
1                                         1001020200                                   
2                                         1001020300                                   
3                                         1001020400                                   
4                                         1001020500                                   

   Primary RUCA Code 2010  Secondary RUCA Code, 2010 (see errata)  \
0                       1                                     1.0   
1                       1                                     1.0   
2                       1                                     1.0   
3                       1                                     1.0   
4                       1                                     1.0   

   Tract Population, 2010  Land Area (square miles), 2010  \
0                    1912                        3.787641   
1                    2170                        1.289776   
2                    3373                        2.065366   
3                    4386                        2.464376   
4                   10766                        4.400686   

   Population Density (per square mile), 2010  
0                                  504.799727  
1                                 1682.462370  
2                                 1633.124331  
3                                 1779.760676  
4                                 2446.436531

In [1972]:
ruca_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74002 entries, 0 to 74001
Data columns (total 9 columns):
 #   Column                                                                             Non-Null Count  Dtype  
---  ------                                                                             --------------  -----  
 0   State-County FIPS Code                                                             74002 non-null  int64  
 1   Select State                                                                       74002 non-null  object 
 2   Select County                                                                      74002 non-null  object 
 3   State-County-Tract FIPS Code (lookup by address at http://www.ffiec.gov/Geocode/)  74002 non-null  int64  
 4   Primary RUCA Code 2010                                                             74002 non-null  int64  
 5   Secondary RUCA Code, 2010 (see errata)                                             74002 non-null  floa

The Primary RUCA code is classified in a range of 1 - 10: 
1	Metropolitan area core: primary flow within an urbanized area (UA),
2	Metropolitan area high commuting: primary flow 30% or more to a UA,
3	Metropolitan area low commuting: primary flow 10% to 30% to a UA,
4	Micropolitan area core: primary flow within an urban cluster of 10,000 to 49,999 (large UC),
5	Micropolitan high commuting: primary flow 30% or more to a large UC,
6	Micropolitan low commuting: primary flow 10% to 30% to a large UC,
7	Small town core: primary flow within an urban cluster of 2,500 to 9,999 (small UC),
8	Small town high commuting: primary flow 30% or more to a small UC,
9	Small town low commuting: primary flow 10% to 30% to a small UC,
10	Rural areas: primary flow to a tract outside a UA or UC,
99	Not coded: Census tract has zero population and no rural-urban identifier information

These classification codes are identified using measures of population density, urbanization, and daily commuting.

Although in 2022, the US Census Bureau's criteria to define urban areas represent a significant departure from previous decades. First, population density was the primary statistic used in the delineation of urban areas from 1960 to 2010. In general, urban areas had a core with a population density of 1,000 persons per square mile and included adjoining territory with at least 500 persons per square mile. In 2020, housing unit density replaced population density as the primary measure of designating census blocks as urban. Second, the minimum population size to qualify as an urban area was 2,500—a threshold that had been in place since 1910. The minimum size for qualification in 2020 was 2,000 housing units or 5,000 people.
Link to the full documentation: https://www.ers.usda.gov/topics/rural-economy-population/rural-classifications/what-is-rural#:~:text=According%20to%20the%20current%20delineation,housing%20units%20and%205%2C000%20residents.

Based on this documentation, the RUCA 2010 dataset we got from the USDA has an outdated classification criteria compared to the criteria mentioned above. The RUCA 2010 data used measures of population density as its primary statistic while in 2020, housing unit density replaced population density as the primary measure. However, the 2020 version of the RUCA has not been released yet. 

The RUCA codes 2020 version according to the USDA will be released no earlier than Spring 2025 (this includes updating its RUCA codes using population data from the 2020 Census and the latest urban area delineations from the Census Bureau, which were released in December 2022. ERS’s RUCA codes also require a third input: updated commuting data from the American Community Survey, measured at the census tract level.)

In [1974]:
# classify the PRIMARY RUCA CODES to either Urban, Suburban and Rural 
# read the ruca code description for more information regarding the classification 

# def classify(ruca_code):
#     if ruca_code == 99:
#         return 'Unknown'
#     elif ruca_code in [1, 2, 3]: 
#         return 'Urban'
#     elif ruca_code in [4, 5, 6]:
#         return 'Suburban'
#     else:
#         return 'Rural'

# feature engineer community type, subsidiving the codes 1-10 into 4 categories: Unknown, Urban, Suburban and Rural
ruca_df['Community Type'] = ruca_df['Primary RUCA Code 2010'].apply(lambda x: 'Unknown' if x == 99 else
                                                                              'Urban' if x in [1, 2, 3] else
                                                                              'Suburban' if x in [4, 5, 6] else
                                                                              'Rural')


In [1975]:
ruca_df['Community Type'].value_counts()

Community Type
Urban       60238
Rural        6796
Suburban     6644
Unknown       324
Name: count, dtype: int64

In [1976]:
# renaming State-County-Tract FIPS Code and using .zfill to make sure FIPS 11 has 11 digits
ruca_df = ruca_df.rename(columns={'State-County-Tract FIPS Code (lookup by address at http://www.ffiec.gov/Geocode/)': 'FIPS 11'})

ruca_df['FIPS 11'] = ruca_df['FIPS 11'].astype(str)
ruca_df['FIPS 11'] = ruca_df['FIPS 11'].apply(lambda x: x.zfill(11))

In [1977]:
ruca_df.head()

State-County FIPS Code Select State   Select County      FIPS 11  \
0                    1001           AL  Autauga County  01001020100   
1                    1001           AL  Autauga County  01001020200   
2                    1001           AL  Autauga County  01001020300   
3                    1001           AL  Autauga County  01001020400   
4                    1001           AL  Autauga County  01001020500   

   Primary RUCA Code 2010  Secondary RUCA Code, 2010 (see errata)  \
0                       1                                     1.0   
1                       1                                     1.0   
2                       1                                     1.0   
3                       1                                     1.0   
4                       1                                     1.0   

   Tract Population, 2010  Land Area (square miles), 2010  \
0                    1912                        3.787641   
1                    2170                        1.289776   
2                    3373                        2.065366   
3                    4386                        2.464376   
4                   10766                        4.400686   

   Population Density (per square mile), 2010 Community Type  
0                                  504.799727          Urban  
1                                 1682.462370          Urban  
2                                 1633.124331          Urban  
3                                 1779.760676          Urban  
4                                 2446.436531          Urban

In [1978]:
ruca_df['FIPS 11'].nunique()

74002

##### State-level 2020 Census Tract to 2010 Census Tract Relationship Files 
- This will help us map FIPS 11 code from 2010 to its 2020 version, this hopefully will help us get better matches once we merge RUCA with the US Census Bureau data

In [1980]:
tract_20_to_10 = pd.read_csv('tab20_tract20_tract10_natl.txt', delimiter = '|')
# tract_20_to_10.to_csv('tract20_tract10.csv', index = False)

In [1981]:
tract20_tract10 = pd.read_csv('tract20_tract10.csv')
tract20_tract10_df = tract20_tract10.copy(deep = True)

In [1982]:
tract20_tract10_df.head()

OID_TRACT_20  GEOID_TRACT_20 NAMELSAD_TRACT_20  AREALAND_TRACT_20  \
0  20790540092527      1001020100  Census Tract 201            9825304   
1  20790540092527      1001020100  Census Tract 201            9825304   
2  20790540092534      1001020200  Census Tract 202            3320818   
3  20790540092528      1001020300  Census Tract 203            5349271   
4  20790540092529      1001020400  Census Tract 204            6384282   

   AREAWATER_TRACT_20 MTFCC_TRACT_20 FUNCSTAT_TRACT_20    OID_TRACT_10  \
0               28435          G5020                 S  20740540092527   
1               28435          G5020                 S  20740540092534   
2                5669          G5020                 S  20740540092534   
3                9054          G5020                 S  20740540092528   
4                8408          G5020                 S  20740540092529   

   GEOID_TRACT_10 NAMELSAD_TRACT_10  AREALAND_TRACT_10  AREAWATER_TRACT_10  \
0      1001020100  Census Tract 201            9827271               28435   
1      1001020200  Census Tract 202            3325674                5669   
2      1001020200  Census Tract 202            3325674                5669   
3      1001020300  Census Tract 203            5349271                9054   
4      1001020400  Census Tract 204            6384282                8408   

  MTFCC_TRACT_10 FUNCSTAT_TRACT_10  AREALAND_PART  AREAWATER_PART  
0          G5020                 S        9820448           28435  
1          G5020                 S           4856               0  
2          G5020                 S        3320818            5669  
3          G5020                 S        5349271            9054  
4          G5020                 S        6384282            8408

In [1983]:
tract20_tract10_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126450 entries, 0 to 126449
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   OID_TRACT_20        126450 non-null  int64 
 1   GEOID_TRACT_20      126450 non-null  int64 
 2   NAMELSAD_TRACT_20   126450 non-null  object
 3   AREALAND_TRACT_20   126450 non-null  int64 
 4   AREAWATER_TRACT_20  126450 non-null  int64 
 5   MTFCC_TRACT_20      126450 non-null  object
 6   FUNCSTAT_TRACT_20   126450 non-null  object
 7   OID_TRACT_10        126450 non-null  int64 
 8   GEOID_TRACT_10      126450 non-null  int64 
 9   NAMELSAD_TRACT_10   126450 non-null  object
 10  AREALAND_TRACT_10   126450 non-null  int64 
 11  AREAWATER_TRACT_10  126450 non-null  int64 
 12  MTFCC_TRACT_10      126450 non-null  object
 13  FUNCSTAT_TRACT_10   126450 non-null  object
 14  AREALAND_PART       126450 non-null  int64 
 15  AREAWATER_PART      126450 non-null  int64 
dtypes:

In [1984]:
# change datatype of GEOID_TRACT_20 AND GEOID_TRACT_10 to str and zfill
tract20_tract10_df['GEOID_TRACT_10'] = tract20_tract10_df['GEOID_TRACT_10'].astype(str).str.zfill(11)
tract20_tract10_df['GEOID_TRACT_20'] = tract20_tract10_df['GEOID_TRACT_20'].astype(str).str.zfill(11)
tract20_tract10_df['AREALAND_TRACT_20'] = tract20_tract10_df['AREALAND_TRACT_20'].astype(float)

In [1985]:
tract20_tract10_df

OID_TRACT_20 GEOID_TRACT_20  NAMELSAD_TRACT_20  AREALAND_TRACT_20  \
0        20790540092527    01001020100   Census Tract 201          9825304.0   
1        20790540092527    01001020100   Census Tract 201          9825304.0   
2        20790540092534    01001020200   Census Tract 202          3320818.0   
3        20790540092528    01001020300   Census Tract 203          5349271.0   
4        20790540092529    01001020400   Census Tract 204          6384282.0   
...                 ...            ...                ...                ...   
126445  207904252102449    78030961100  Census Tract 9611          3479232.0   
126446   20790228304757    78030961200  Census Tract 9612          1101324.0   
126447   20790228304757    78030961200  Census Tract 9612          1101324.0   
126448   20790228304757    78030961200  Census Tract 9612          1101324.0   
126449   20790228304801    78030990000  Census Tract 9900                0.0   

        AREAWATER_TRACT_20 MTFCC_TRACT_20 FUNCSTAT_TRACT_20     OID_TRACT_10  \
0                    28435          G5020                 S   20740540092527   
1                    28435          G5020                 S   20740540092534   
2                     5669          G5020                 S   20740540092534   
3                     9054          G5020                 S   20740540092528   
4                     8408          G5020                 S   20740540092529   
...                    ...            ...               ...              ...   
126445                   0          G5020                 S   20740228304757   
126446              802134          G5020                 S   20740228304707   
126447              802134          G5020                 S  207404252102449   
126448              802134          G5020                 S   20740228304757   
126449           687337537          G5020                 S   20740228304801   

       GEOID_TRACT_10  NAMELSAD_TRACT_10  AREALAND_TRACT_10  \
0         01001020100   Census Tract 201            9827271   
1         01001020200   Census Tract 202            3325674   
2         01001020200   Census Tract 202            3325674   
3         01001020300   Census Tract 203            5349271   
4         01001020400   Census Tract 204            6384282   
...               ...                ...                ...   
126445    78030961200  Census Tract 9612            1017540   
126446    78030960400  Census Tract 9604           11709358   
126447    78030961100  Census Tract 9611            3513895   
126448    78030961200  Census Tract 9612            1017540   
126449    78030990000  Census Tract 9900                  0   

        AREAWATER_TRACT_10 MTFCC_TRACT_10 FUNCSTAT_TRACT_10  AREALAND_PART  \
0                    28435          G5020                 S        9820448   
1                     5669          G5020                 S           4856   
2                     5669          G5020                 S        3320818   
3                     9054          G5020                 S        5349271   
4                     8408          G5020                 S        6384282   
...                    ...            ...               ...            ...   
126445              802134          G5020                 S          14104   
126446              413661          G5020                 S         217781   
126447                   0          G5020                 S            203   
126448              802134          G5020                 S         883340   
126449           687428211          G5020                 S              0   

        AREAWATER_PART  
0                28435  
1                    0  
2                 5669  
3                 9054  
4                 8408  
...                ...  
126445               0  
126446               0  
126447               0  
126448          802134  
126449       687337537  

[126450 rows x 16 columns]

In [1986]:
tract20_tract10_df['GEOID_TRACT_10'].nunique()

74134

In [1987]:
tract20_tract10_df['GEOID_TRACT_10'].duplicated().sum()

52316

In [1988]:
# 74002 matches of FIPS 11 from RUCA is in GEOID_TRACT_10
matching = ruca_df[ruca_df['FIPS 11'].isin(tract20_tract10_df['GEOID_TRACT_10'])]
print(matching.shape)

(74002, 10)


In [1989]:
# 94287 of the GEOID_TRACT_20 is in GEOID_TRACT_10, the differece (126450 - 94287 rows) shows that there are definitely changes in some FIPS code from the 2010 to 2020 version
matching1 = tract20_tract10_df[tract20_tract10_df['GEOID_TRACT_20'].isin(tract20_tract10_df['GEOID_TRACT_10'])]
matching1.shape

(94287, 16)

In [1990]:
# since we only need the GEOID_TRACT_20 and GEOID_TRACT_10, let's separate it from the full tract20_tract10_df
tr20_tr10 = tract20_tract10_df[['GEOID_TRACT_20', 'GEOID_TRACT_10', 'AREALAND_TRACT_20']].drop_duplicates(subset='GEOID_TRACT_10')
tr20_tr10.rename(columns = {'GEOID_TRACT_10': 'FIPS 11'}, inplace = True)

In [1991]:
# AREALAND_TRACT_2020 will be land area in square miles 
tr20_tr10['AREALAND_TRACT_20'] = tract20_tract10_df['AREALAND_TRACT_20'] / 1000000

In [1992]:
# # filter only relevant FIPS 11 values from tr20_tr10 that are in ruca df
# tr20_tr10_filtered = tr20_tr10[tr20_tr10['FIPS 11'].isin(ruca_df['FIPS 11'])]

# # frop duplicate FIPS 11 values (keeping only one row per FIPS 11)
# tr20_tr10_filtered = tr20_tr10_filtered.drop_duplicates(subset=['FIPS 11'])

# # merge with RUCA data
# ruca_updated = pd.merge(ruca_df, tr20_tr10_filtered, on='FIPS 11', how='inner')

ruca_updated = pd.merge(ruca_df, tr20_tr10, on='FIPS 11', how='inner')

In [2137]:
# this ruca_updated dataframe will be merged with the US Census Bureau since both FIPS codes are now using 2020 version
ruca_updated.shape

(74002, 12)

In [2135]:
ruca_updated['GEOID_TRACT_20'].nunique()

62212

### NCUA data 
- The NCUA is responsible for regulating credit unions
- The NCUA dataset contains data regarding credit unions in the US

In [1995]:
# ncua_df_orig = pd.read_csv('ncua/combined_file.csv')
# ncua_df_ = ncua.copy(deep = True)

In [1996]:
# ncua_ df 
# ncua_df.head()

In [1997]:
# ncua_df.info()

In [1998]:
# dropping columns not needed 
# ncua_df.drop(columns = ['Charter number', 'Year and quarter', 'Credit Union type', 'NCUA region', 'Net worth ratio (excludes CECL transition provision)', 'Loan-to-share ratio', 'Total deposits,4 quarter growth (%)', 'Total loans, 4 quarter growth (%)', 'Total assets, 4 quarter growth (%)', 'Members,4 quarter growth (%)', 'Net worth, 4 quarter growth (excludes CECL transition provision)(%)', 'Unnamed: 24', 'Match Type', 'Match Precision', 'Standardized Address', 'Unknown1'], inplace = True)

In [1999]:
# check null values 
# ncua_df.isna().sum()

In [2000]:
# dropping null values
# ncua_df.dropna(inplace = True)
# ncua_df.reset_index(drop = True)

In [2001]:
# change data type of state, county, tract and block code to str
# ncua_df['State Code'] = ncua_df['State Code'].astype(str)
# ncua_df['County Code'] = ncua_df['County Code'].astype(str)
# ncua_df['Tract Code'] = ncua_df['Tract Code'].astype(str)
# ncua_df['Block Code'] = ncua_df['Block Code'].astype(str)

# # filling the tract code columns with zeroes until it's 6 digits long 
# ncua_df['Tract Code'] = ncua_df['Tract Code'].apply(lambda x: x.zfill(6))
# ncua_df['State Code'] = ncua_df['State Code'].apply(lambda x: x.zfill(2))
# ncua_df['County Code'] = ncua_df['County Code'].apply(lambda x: x.zfill(3))

In [2002]:
# ncua_df.head()

After dropping all null values in the NCUA dataset we are left with 2761 credit unions 

### Feature Engineering geocoded addresses from NCUA dataset

In [2005]:
# Longitude column
# ncua_df['Longitude'] = ncua_df['Coordinates'].apply(lambda x: x.split(',')[0])

# # Latitude column
# ncua_df['Latitude'] = ncua_df['Coordinates'].apply(lambda x: x.split(',')[1])

In [2006]:
# feature engineer a FULL 11 DIGIT FIPS CODE (state, county, tract code)

# ncua_df['FIPS 11'] = ncua_df.apply(lambda row: row['State Code'] + row['County Code'] + row['Tract Code'], axis = 1)

In [2007]:
# ncua_df

### US Census Bureau 

In [2009]:
census = pd.read_csv('us_census/Census_ACS5_Export.csv')
census_df = census.copy(deep = True)

In [2010]:
census_df.tail()

GEO_ID                                         NAME   Pop  \
84410  56043000200     Census Tract 2; Washakie County; Wyoming  3028   
84411  56043000301  Census Tract 3.01; Washakie County; Wyoming  2323   
84412  56043000302  Census Tract 3.02; Washakie County; Wyoming  2374   
84413  56045951100    Census Tract 9511; Weston County; Wyoming  3275   
84414  56045951300    Census Tract 9513; Weston County; Wyoming  3595   

       Hous_Units   Income  Poverty%    White%    Black%  Hispanic%    Asian%  \
84410        1580  56627.0  0.065390  0.910502  0.000330   0.063738  0.010568   
84411        1128  61087.0  0.038743  0.832975  0.000000   0.247094  0.007318   
84412        1148  72634.0  0.090564  0.822241  0.000000   0.141533  0.003370   
84413        1557  67715.0  0.140153  0.838168  0.001832   0.025954  0.022901   
84414        1855  77644.0  0.103199  0.888456  0.000000   0.043950  0.023088   

       TwoOrMore%     Male%   Female%  Under18%      65+%  Employment%  \
84410    0.026750  0.532034  0.467966  0.206407  0.211361     0.981280   
84411    0.144210  0.565217  0.434783  0.289712  0.184675     0.995675   
84412    0.120472  0.472199  0.527801  0.165122  0.308762     0.970516   
84413    0.130382  0.551756  0.448244  0.170076  0.254962     0.946086   
84414    0.068985  0.531572  0.468428  0.235327  0.165508     0.977839   

       Disability%  VacantHous%  OwnOcpHous%  
84410     0.105020     0.134810     0.644937  
84411     0.130435     0.159574     0.542553  
84412     0.170598     0.042683     0.709930  
84413     0.139542     0.118176     0.806680  
84414     0.122392     0.318059     0.603235

The GEOID column needs to be exactly 11 digits so we need to add 0 in the beginning for GEOID's with less than 11 digits

In [2012]:
census_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84415 entries, 0 to 84414
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   GEO_ID       84415 non-null  int64  
 1   NAME         84415 non-null  object 
 2   Pop          84415 non-null  int64  
 3   Hous_Units   84415 non-null  int64  
 4   Income       82898 non-null  float64
 5   Poverty%     83617 non-null  float64
 6   White%       83617 non-null  float64
 7   Black%       83617 non-null  float64
 8   Hispanic%    83617 non-null  float64
 9   Asian%       83617 non-null  float64
 10  TwoOrMore%   83617 non-null  float64
 11  Male%        83617 non-null  float64
 12  Female%      83617 non-null  float64
 13  Under18%     83617 non-null  float64
 14  65+%         83617 non-null  float64
 15  Employment%  83477 non-null  float64
 16  Disability%  83617 non-null  float64
 17  VacantHous%  83388 non-null  float64
 18  OwnOcpHous%  83388 non-null  float64
dtypes: f

In [2013]:
census_df.describe()

GEO_ID           Pop    Hous_Units         Income      Poverty%  \
count  8.441500e+04  84415.000000  84415.000000   82898.000000  83617.000000   
mean   2.786884e+10   3922.260179   1669.651282   80716.700765      0.130821   
std    1.593237e+10   1744.017379    705.692595   39409.311162      0.105462   
min    1.001020e+09      0.000000      0.000000    2499.000000      0.000000   
25%    1.212197e+10   2706.000000   1196.000000   53750.000000      0.054935   
50%    2.714103e+10   3754.000000   1609.000000   71944.000000      0.102294   
75%    4.106703e+10   4941.000000   2077.000000   98454.500000      0.177195   
max    5.604595e+10  38907.000000  12980.000000  250001.000000      1.000000   

             White%        Black%     Hispanic%        Asian%    TwoOrMore%  \
count  83617.000000  83617.000000  83617.000000  83617.000000  83617.000000   
mean       0.662136      0.132994      0.175005      0.009232      0.084317   
std        0.259157      0.208329      0.218495      0.046219      0.075840   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.493290      0.008467      0.030493      0.000000      0.032233   
50%        0.730182      0.042040      0.084721      0.000000      0.063525   
75%        0.878669      0.154196      0.224014      0.005661      0.112107   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

              Male%       Female%      Under18%          65+%   Employment%  \
count  83617.000000  83617.000000  83617.000000  83617.000000  83477.000000   
mean       0.496515      0.503485      0.214038      0.173008      0.944024   
std        0.053853      0.053853      0.072190      0.090214      0.045833   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.469002      0.478739      0.173342      0.116718      0.926488   
50%        0.495642      0.504358      0.215836      0.163276      0.954749   
75%        0.521261      0.530998      0.257701      0.213305      0.974302   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

        Disability%   VacantHous%   OwnOcpHous%  
count  83617.000000  83388.000000  83388.000000  
mean       0.133078      0.107280      0.578544  
std        0.061819      0.106980      0.227573  
min        0.000000      0.000000      0.000000  
25%        0.089647      0.039134      0.428707  
50%        0.124351      0.077649      0.610199  
75%        0.167397      0.139249      0.753892  
max        1.000000      1.000000      1.000000

In [2014]:
census_df['GEO_ID'].nunique()

84415

In [2015]:
# check nan values 
census_df.isna().sum()

GEO_ID            0
NAME              0
Pop               0
Hous_Units        0
Income         1517
Poverty%        798
White%          798
Black%          798
Hispanic%       798
Asian%          798
TwoOrMore%      798
Male%           798
Female%         798
Under18%        798
65+%            798
Employment%     938
Disability%     798
VacantHous%    1027
OwnOcpHous%    1027
dtype: int64

In [2016]:
# drop all null values 
census_df.dropna(inplace = True)
census_df.reset_index(drop = True)

GEO_ID                                          NAME   Pop  \
0       1001020100     Census Tract 201; Autauga County; Alabama  1865   
1       1001020200     Census Tract 202; Autauga County; Alabama  1861   
2       1001020300     Census Tract 203; Autauga County; Alabama  3492   
3       1001020400     Census Tract 204; Autauga County; Alabama  3987   
4       1001020501  Census Tract 205.01; Autauga County; Alabama  4121   
...            ...                                           ...   ...   
82890  56043000200      Census Tract 2; Washakie County; Wyoming  3028   
82891  56043000301   Census Tract 3.01; Washakie County; Wyoming  2323   
82892  56043000302   Census Tract 3.02; Washakie County; Wyoming  2374   
82893  56045951100     Census Tract 9511; Weston County; Wyoming  3275   
82894  56045951300     Census Tract 9513; Weston County; Wyoming  3595   

       Hous_Units   Income  Poverty%    White%    Black%  Hispanic%    Asian%  \
0             733  60563.0  0.153351  0.804290  0.111528   0.043432  0.000000   
1             680  57460.0  0.056421  0.363246  0.559914   0.001075  0.000000   
2            1431  77371.0  0.100802  0.691008  0.250859   0.012600  0.000000   
3            1722  73191.0  0.102333  0.879609  0.074492   0.011036  0.001756   
4            1857  79953.0  0.078379  0.796166  0.150449   0.036884  0.000000   
...           ...      ...       ...       ...       ...        ...       ...   
82890        1580  56627.0  0.065390  0.910502  0.000330   0.063738  0.010568   
82891        1128  61087.0  0.038743  0.832975  0.000000   0.247094  0.007318   
82892        1148  72634.0  0.090564  0.822241  0.000000   0.141533  0.003370   
82893        1557  67715.0  0.140153  0.838168  0.001832   0.025954  0.022901   
82894        1855  77644.0  0.103199  0.888456  0.000000   0.043950  0.023088   

       TwoOrMore%     Male%   Female%  Under18%      65+%  Employment%  \
0        0.055228  0.519035  0.480965  0.239678  0.194638     0.974255   
1        0.072542  0.544331  0.455669  0.098872  0.157442     0.946146   
2        0.048110  0.460195  0.539805  0.261741  0.134593     0.981748   
3        0.038375  0.510158  0.489842  0.145974  0.260095     0.990933   
4        0.029847  0.465664  0.534336  0.181267  0.133705     0.994409   
...           ...       ...       ...       ...       ...          ...   
82890    0.026750  0.532034  0.467966  0.206407  0.211361     0.981280   
82891    0.144210  0.565217  0.434783  0.289712  0.184675     0.995675   
82892    0.120472  0.472199  0.527801  0.165122  0.308762     0.970516   
82893    0.130382  0.551756  0.448244  0.170076  0.254962     0.946086   
82894    0.068985  0.531572  0.468428  0.235327  0.165508     0.977839   

       Disability%  VacantHous%  OwnOcpHous%  
0         0.200000     0.045020     0.708049  
1         0.180548     0.200000     0.630882  
2         0.158362     0.088050     0.637317  
3         0.222222     0.032520     0.758420  
4         0.173016     0.039849     0.522886  
...            ...          ...          ...  
82890     0.105020     0.134810     0.644937  
82891     0.130435     0.159574     0.542553  
82892     0.170598     0.042683     0.709930  
82893     0.139542     0.118176     0.806680  
82894     0.122392     0.318059     0.603235  

[82895 rows x 19 columns]

In [2017]:
# create majority group, majority percentage, minority group and minority percentage 
# define race columns
race_columns = ["White%", "Black%", "Hispanic%", "Asian%", "TwoOrMore%"]

# find the majority race (highest percentage)
census_df["majority_race"] = census_df[race_columns].idxmax(axis=1)
census_df["majority_race_percentage"] = census_df[race_columns].max(axis=1)

# find the minority race (second highest percentage)
census_df["minority_race"] = census_df[race_columns].apply(lambda row: row.nlargest(2).idxmin(), axis=1)
census_df["minority_race_percentage"] = census_df[race_columns].apply(lambda row: row.nlargest(2).min(), axis=1)

# Show updated DataFrame
census_df.head()

GEO_ID                                          NAME   Pop  Hous_Units  \
0  1001020100     Census Tract 201; Autauga County; Alabama  1865         733   
1  1001020200     Census Tract 202; Autauga County; Alabama  1861         680   
2  1001020300     Census Tract 203; Autauga County; Alabama  3492        1431   
3  1001020400     Census Tract 204; Autauga County; Alabama  3987        1722   
4  1001020501  Census Tract 205.01; Autauga County; Alabama  4121        1857   

    Income  Poverty%    White%    Black%  Hispanic%    Asian%  TwoOrMore%  \
0  60563.0  0.153351  0.804290  0.111528   0.043432  0.000000    0.055228   
1  57460.0  0.056421  0.363246  0.559914   0.001075  0.000000    0.072542   
2  77371.0  0.100802  0.691008  0.250859   0.012600  0.000000    0.048110   
3  73191.0  0.102333  0.879609  0.074492   0.011036  0.001756    0.038375   
4  79953.0  0.078379  0.796166  0.150449   0.036884  0.000000    0.029847   

      Male%   Female%  Under18%      65+%  Employment%  Disability%  \
0  0.519035  0.480965  0.239678  0.194638     0.974255     0.200000   
1  0.544331  0.455669  0.098872  0.157442     0.946146     0.180548   
2  0.460195  0.539805  0.261741  0.134593     0.981748     0.158362   
3  0.510158  0.489842  0.145974  0.260095     0.990933     0.222222   
4  0.465664  0.534336  0.181267  0.133705     0.994409     0.173016   

   VacantHous%  OwnOcpHous% majority_race  majority_race_percentage  \
0     0.045020     0.708049        White%                  0.804290   
1     0.200000     0.630882        Black%                  0.559914   
2     0.088050     0.637317        White%                  0.691008   
3     0.032520     0.758420        White%                  0.879609   
4     0.039849     0.522886        White%                  0.796166   

  minority_race  minority_race_percentage  
0        Black%                  0.111528  
1        White%                  0.363246  
2        Black%                  0.250859  
3        Black%                  0.074492  
4        Black%                  0.150449

Summarized all races White%, Black%, Asian%, Hispanic%, TwoOrMore% into majority race and minority race with the percentages under majority/minority percentages column. 

In [2019]:
# define gender columns 
gender_columns = ['Male%', 'Female%']

# find majority gender (highest percentage) 
census_df["majority_gender"] = census_df[gender_columns].idxmax(axis=1)
census_df["majority_gender_percentage"] = census_df[gender_columns].max(axis=1)

# find the minority gender (second highest percentage)
census_df["minority_gender"] = census_df[gender_columns].apply(lambda row: row.nlargest(2).idxmin(), axis=1)
census_df["minority_gender_percentage"] = census_df[gender_columns].apply(lambda row: row.nlargest(2).min(), axis=1)

In [2020]:
# create a 18-64% age range 
census_df["18_64%"] = 1 - (census_df["Under18%"] + census_df["65+%"])

In [2021]:
# define age columns 
age_columns = ['Under18%', '18_64%', '65+%']

# find majority age (highest percentage) 
census_df["majority_age"] = census_df[age_columns].idxmax(axis=1)
census_df["majority_age_percentage"] = census_df[age_columns].max(axis=1)

# find the minority age (second highest percentage)
census_df["minority_age"] = census_df[age_columns].apply(lambda row: row.nlargest(2).idxmin(), axis=1)
census_df["minority_age_percentage"] = census_df[age_columns].apply(lambda row: row.nlargest(2).min(), axis=1)

In [2022]:
# dropping unnecessary columns 
census_df.drop(columns = ["White%", "Black%", "Hispanic%", "Asian%", "TwoOrMore%", "Male%", "Female%", "Under18%", "18_64%", "65+%"], inplace = True)

In [2023]:
# GEO_ID is the census tracts 11 digit fips code 
# in order for GEO_ID to match with the other datasets, we have to make sure it has 11 digits 
# change datatype from int to str so we can use .zfill() 

census_df['GEO_ID'] = census_df['GEO_ID'].astype(str)

# filling the tract code columns with zeroes until it's 6 digits long 
census_df['GEO_ID'] = census_df['GEO_ID'].apply(lambda x: x.zfill(11))

In [2024]:
# GEO_ID is the same as FIPS 11 -- it contains state-county-tract 11 digit code
# rename GEO_ID to FIPS 11 for uniformity 

census_df = census_df.rename(columns = {'GEO_ID': 'GEOID_TRACT_20', # use GEOID_TRACT_20, FIPS 11 is for the 2020 version
                                       'NAME': 'Census Tract Name',
                                       'Pop': 'Population', 
                                       })

In [2025]:
census_df.head()

GEOID_TRACT_20                             Census Tract Name  Population  \
0    01001020100     Census Tract 201; Autauga County; Alabama        1865   
1    01001020200     Census Tract 202; Autauga County; Alabama        1861   
2    01001020300     Census Tract 203; Autauga County; Alabama        3492   
3    01001020400     Census Tract 204; Autauga County; Alabama        3987   
4    01001020501  Census Tract 205.01; Autauga County; Alabama        4121   

   Hous_Units   Income  Poverty%  Employment%  Disability%  VacantHous%  \
0         733  60563.0  0.153351     0.974255     0.200000     0.045020   
1         680  57460.0  0.056421     0.946146     0.180548     0.200000   
2        1431  77371.0  0.100802     0.981748     0.158362     0.088050   
3        1722  73191.0  0.102333     0.990933     0.222222     0.032520   
4        1857  79953.0  0.078379     0.994409     0.173016     0.039849   

   OwnOcpHous% majority_race  majority_race_percentage minority_race  \
0     0.708049        White%                  0.804290        Black%   
1     0.630882        Black%                  0.559914        White%   
2     0.637317        White%                  0.691008        Black%   
3     0.758420        White%                  0.879609        Black%   
4     0.522886        White%                  0.796166        Black%   

   minority_race_percentage majority_gender  majority_gender_percentage  \
0                  0.111528           Male%                    0.519035   
1                  0.363246           Male%                    0.544331   
2                  0.250859         Female%                    0.539805   
3                  0.074492           Male%                    0.510158   
4                  0.150449         Female%                    0.534336   

  minority_gender  minority_gender_percentage majority_age  \
0         Female%                    0.480965       18_64%   
1         Female%                    0.455669       18_64%   
2           Male%                    0.460195       18_64%   
3         Female%                    0.489842       18_64%   
4           Male%                    0.465664       18_64%   

   majority_age_percentage minority_age  minority_age_percentage  
0                 0.565684     Under18%                 0.239678  
1                 0.743686         65+%                 0.157442  
2                 0.603666     Under18%                 0.261741  
3                 0.593930         65+%                 0.260095  
4                 0.685028     Under18%                 0.181267

In [2133]:
census_df['GEOID_TRACT_20'].nunique()

82895

The number of unique fips 11 (census tract code) from the RUCA dataset contains 74002 unique fips, while the US Census Bureau dataset contains 82895. We will only be keeping rows that are matching in both datasets.

### Merging RUCA dataset and US census dataset
- The discrepancy between the FIPS 11 from RUCA and FIPS 11 from the US Census Bureau is that they are both from different versions of datasets. The RUCA code is a 2010 version while the US Census data is from the ACS5 2020
- Since the RUCA code is from the 2010 version, the FIPS code it's using might have changed
- We can use the state level 2020 Census Tract to 2010 Census Tract Relationship Files to map the old RUCA fips 11 to the 2020 version so it matches better with the US Census Bureau data
- More details regarding the 2020 Census Tract to 2010 Census Tract Relationship Files: https://www.census.gov/geographies/reference-files/time-series/geo/relationship-files.2020.html#tract
- ruca_updated dataframe contains both fips 11 2010 and 2020 version, the 2020 version is named GEOID_TRACT_20

In [2028]:
ruca_census_df = pd.merge(ruca_updated, census_df, on = 'GEOID_TRACT_20', how = 'inner')

##### The merging of updated RUCA data with 2020 version of FIPS 11 had more matches with the US Census Bureau 2020 ACS5, resulting in 71275 rows.

In [2030]:
# drop unnecessary and redundant columns 
ruca_census_df.drop(columns = ['State-County FIPS Code', 'Secondary RUCA Code, 2010 (see errata)', 'Tract Population, 2010', 'Land Area (square miles), 2010', 'Population Density (per square mile), 2010'], inplace = True)

In [2031]:
# rename columns 
ruca_census_df.rename(columns = {
    'Select State': 'State Code', 
    'Select County': 'County Code', 
    'FIPS 11': 'Geoid Tract 2010',
    'GEOID_TRACT_20': 'Geoid Tract 2020', 
    'AREALAND_TRACT_20': 'Land Area (in Sq. Miles) 2020', 
    'majority_race': 'Majority Race', 
    'majority_race_percentage': 'Majority Race Percentage', 
    'minority_race': 'Minority Race', 
    'minority_race_percentage': 'Minority Race Percentage', 
    'majority_gender': 'Majority Gender', 
    'majority_gender_percentage': 'Majority Gender Percentage', 
    'minority_gender': 'Minority Gender', 
    'minority_gender_percentage': 'Minority Gender Percentage', 
    'majority_age': 'Majority Age', 
    'majority_age_percentage': 'Majority Age Percentage', 
    'minority_age': 'Minority Age', 
    'minority_age_percentage': 'Minority Age Percentage'
}, inplace = True)

##### Feature engineering population and housing density for 2020 
- Population density pertains to the number of people per square mile
- Housing density pertains to the number of housing units per square mile 

In [2033]:
# Feature engineering population and housing density for 2020 
ruca_census_df['Population Density'] = ruca_census_df.apply(lambda row: row['Population'] / row['Land Area (in Sq. Miles) 2020'], axis = 1)
ruca_census_df['Housing Density'] = ruca_census_df.apply(lambda row: row['Hous_Units'] / row['Land Area (in Sq. Miles) 2020'], axis = 1)

In [2034]:
ruca_census_df.head()

State Code     County Code Geoid Tract 2010  Primary RUCA Code 2010  \
0         AL  Autauga County      01001020100                       1   
1         AL  Autauga County      01001020200                       1   
2         AL  Autauga County      01001020300                       1   
3         AL  Autauga County      01001020400                       1   
4         AL  Autauga County      01001020500                       1   

  Community Type Geoid Tract 2020  Land Area (in Sq. Miles) 2020  \
0          Urban      01001020100                       9.825304   
1          Urban      01001020100                       9.825304   
2          Urban      01001020300                       5.349271   
3          Urban      01001020400                       6.384282   
4          Urban      01001020501                       6.203654   

                              Census Tract Name  Population  Hous_Units  \
0     Census Tract 201; Autauga County; Alabama        1865         733   
1     Census Tract 201; Autauga County; Alabama        1865         733   
2     Census Tract 203; Autauga County; Alabama        3492        1431   
3     Census Tract 204; Autauga County; Alabama        3987        1722   
4  Census Tract 205.01; Autauga County; Alabama        4121        1857   

    Income  Poverty%  Employment%  Disability%  VacantHous%  OwnOcpHous%  \
0  60563.0  0.153351     0.974255     0.200000     0.045020     0.708049   
1  60563.0  0.153351     0.974255     0.200000     0.045020     0.708049   
2  77371.0  0.100802     0.981748     0.158362     0.088050     0.637317   
3  73191.0  0.102333     0.990933     0.222222     0.032520     0.758420   
4  79953.0  0.078379     0.994409     0.173016     0.039849     0.522886   

  Majority Race  Majority Race Percentage Minority Race  \
0        White%                  0.804290        Black%   
1        White%                  0.804290        Black%   
2        White%                  0.691008        Black%   
3        White%                  0.879609        Black%   
4        White%                  0.796166        Black%   

   Minority Race Percentage Majority Gender  Majority Gender Percentage  \
0                  0.111528           Male%                    0.519035   
1                  0.111528           Male%                    0.519035   
2                  0.250859         Female%                    0.539805   
3                  0.074492           Male%                    0.510158   
4                  0.150449         Female%                    0.534336   

  Minority Gender  Minority Gender Percentage Majority Age  \
0         Female%                    0.480965       18_64%   
1         Female%                    0.480965       18_64%   
2           Male%                    0.460195       18_64%   
3         Female%                    0.489842       18_64%   
4           Male%                    0.465664       18_64%   

   Majority Age Percentage Minority Age  Minority Age Percentage  \
0                 0.565684     Under18%                 0.239678   
1                 0.565684     Under18%                 0.239678   
2                 0.603666     Under18%                 0.261741   
3                 0.593930         65+%                 0.260095   
4                 0.685028     Under18%                 0.181267   

   Population Density  Housing Density  
0          189.816010        74.603290  
1          189.816010        74.603290  
2          652.799232       267.513087  
3          624.502489       269.724928  
4          664.285919       299.339712

In [2035]:
ruca_census_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71275 entries, 0 to 71274
Data columns (total 30 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   State Code                     71275 non-null  object 
 1   County Code                    71275 non-null  object 
 2   Geoid Tract 2010               71275 non-null  object 
 3   Primary RUCA Code 2010         71275 non-null  int64  
 4   Community Type                 71275 non-null  object 
 5   Geoid Tract 2020               71275 non-null  object 
 6   Land Area (in Sq. Miles) 2020  71275 non-null  float64
 7   Census Tract Name              71275 non-null  object 
 8   Population                     71275 non-null  int64  
 9   Hous_Units                     71275 non-null  int64  
 10  Income                         71275 non-null  float64
 11  Poverty%                       71275 non-null  float64
 12  Employment%                    71275 non-null 

In [2036]:
# five summary statistic 
ruca_census_df.describe()

Primary RUCA Code 2010  Land Area (in Sq. Miles) 2020    Population  \
count            71275.000000                   71275.000000  71275.000000   
mean                 2.192648                     126.271996   4007.058632   
std                  3.350660                    1512.788531   1703.650246   
min                  1.000000                       0.022158     17.000000   
25%                  1.000000                       1.696562   2772.000000   
50%                  1.000000                       4.415562   3832.000000   
75%                  2.000000                      30.708094   5015.000000   
max                 99.000000                  221585.760400  38907.000000   

         Hous_Units         Income      Poverty%   Employment%   Disability%  \
count  71275.000000   71275.000000  71275.000000  71275.000000  71275.000000   
mean    1715.086566   79839.118892      0.133256      0.943222      0.134847   
std      687.524332   39355.116122      0.104523      0.044783      0.059748   
min       10.000000    2499.000000      0.000000      0.383865      0.000000   
25%     1229.000000   53083.000000      0.057716      0.925784      0.091998   
50%     1645.000000   70938.000000      0.104999      0.953927      0.126138   
75%     2113.000000   97124.000000      0.180080      0.973289      0.168798   
max    12980.000000  250001.000000      0.857309      1.000000      0.792605   

        VacantHous%   OwnOcpHous%  Majority Race Percentage  \
count  71275.000000  71275.000000              71275.000000   
mean       0.108083      0.576394                  0.746593   
std        0.104432      0.223795                  0.175398   
min        0.000000      0.000000                  0.035923   
25%        0.040576      0.427848                  0.623836   
50%        0.078967      0.606735                  0.778343   
75%        0.140591      0.748416                  0.895296   
max        0.984976      1.000000                  1.000000   

       Minority Race Percentage  Majority Gender Percentage  \
count              71275.000000                71275.000000   
mean                   0.185316                    0.533569   
std                    0.145847                    0.032295   
min                    0.000000                    0.500000   
25%                    0.065883                    0.511823   
50%                    0.142924                    0.525541   
75%                    0.279772                    0.545575   
max                    0.986452                    0.994121   

       Minority Gender Percentage  Majority Age Percentage  \
count                71275.000000             71275.000000   
mean                     0.466431                 0.614167   
std                      0.032295                 0.077691   
min                      0.005879                 0.376271   
25%                      0.454425                 0.565747   
50%                      0.474459                 0.606168   
75%                      0.488177                 0.650553   
max                      0.500000                 1.000000   

       Minority Age Percentage  Population Density  Housing Density  
count             71275.000000        71275.000000     71275.000000  
mean                  0.242768         2135.652973       933.379479  
std                   0.061137         4714.798091      2265.257309  
min                   0.000000            0.011567         0.003646  
25%                   0.207858          128.806697        54.453917  
50%                   0.238751          902.865828       385.954491  
75%                   0.275110         2127.068320       890.140875  
max                   0.492057       102265.547432     74690.856576

In [2037]:
ruca_census_df.describe(include = 'object')

State Code         County Code Geoid Tract 2010 Community Type  \
count       71275               71275            71275          71275   
unique         50                1871            71275              4   
top            CA  Los Angeles County      01001020100          Urban   
freq         7969                2310                1          58150   

       Geoid Tract 2020                              Census Tract Name  \
count             71275                                          71275   
unique            59873                                          59873   
top         06001400100  Census Tract 4001; Alameda County; California   
freq                  8                                              8   

       Majority Race Minority Race Majority Gender Minority Gender  \
count          71275         71275           71275           71275   
unique             5             5               2               2   
top           White%     Hispanic%         Female%           Male%   
freq           55641         28330           38734           38847   

       Majority Age Minority Age  
count         71275        71275  
unique            3            3  
top          18_64%     Under18%  
freq          70290        47489

In [2038]:
# there are multiple instances of geoid tract 2020 that got duplicated 
ruca_census_df['Geoid Tract 2020'].value_counts()

Geoid Tract 2020
06001400100    8
51059492201    8
55041950100    7
27019090702    7
48439100601    7
              ..
20043020100    1
20043020200    1
20043020300    1
20045000301    1
56045951300    1
Name: count, Length: 59873, dtype: int64

In [2143]:
ruca_census_unique = ruca_census_df.drop_duplicates(subset='Geoid Tract 2020')

In [2145]:
ruca_census_unique

State Code      County Code Geoid Tract 2010  Primary RUCA Code 2010  \
0             AL   Autauga County      01001020100                       1   
2             AL   Autauga County      01001020300                       1   
3             AL   Autauga County      01001020400                       1   
4             AL   Autauga County      01001020500                       1   
5             AL   Autauga County      01001020600                       1   
...          ...              ...              ...                     ...   
71268         WY     Uinta County      56041975300                       4   
71270         WY  Washakie County      56043000200                       8   
71271         WY  Washakie County      56043000301                       7   
71273         WY    Weston County      56045951100                      10   
71274         WY    Weston County      56045951300                       7   

      Community Type Geoid Tract 2020  Land Area (in Sq. Miles) 2020  \
0              Urban      01001020100                       9.825304   
2              Urban      01001020300                       5.349271   
3              Urban      01001020400                       6.384282   
4              Urban      01001020501                       6.203654   
5              Urban      01001020600                       8.041611   
...              ...              ...                            ...   
71268       Suburban      56041975301                    2281.714868   
71270          Rural      56043000200                    5780.730055   
71271          Rural      56043000301                       1.993200   
71273          Rural      56045951100                    6100.000349   
71274          Rural      56045951300                     110.803770   

                                  Census Tract Name  Population  Hous_Units  \
0         Census Tract 201; Autauga County; Alabama        1865         733   
2         Census Tract 203; Autauga County; Alabama        3492        1431   
3         Census Tract 204; Autauga County; Alabama        3987        1722   
4      Census Tract 205.01; Autauga County; Alabama        4121        1857   
5         Census Tract 206; Autauga County; Alabama        3839        1551   
...                                             ...         ...         ...   
71268   Census Tract 9753.01; Uinta County; Wyoming        2502        1267   
71270      Census Tract 2; Washakie County; Wyoming        3028        1580   
71271   Census Tract 3.01; Washakie County; Wyoming        2323        1128   
71273     Census Tract 9511; Weston County; Wyoming        3275        1557   
71274     Census Tract 9513; Weston County; Wyoming        3595        1855   

        Income  Poverty%  Employment%  Disability%  VacantHous%  OwnOcpHous%  \
0      60563.0  0.153351     0.974255     0.200000     0.045020     0.708049   
2      77371.0  0.100802     0.981748     0.158362     0.088050     0.637317   
3      73191.0  0.102333     0.990933     0.222222     0.032520     0.758420   
4      79953.0  0.078379     0.994409     0.173016     0.039849     0.522886   
5      64904.0  0.151081     0.969313     0.158896     0.135397     0.700193   
...        ...       ...          ...          ...          ...          ...   
71268  73096.0  0.105516     1.000000     0.213829     0.211523     0.644830   
71270  56627.0  0.065390     0.981280     0.105020     0.134810     0.644937   
71271  61087.0  0.038743     0.995675     0.130435     0.159574     0.542553   
71273  67715.0  0.140153     0.946086     0.139542     0.118176     0.806680   
71274  77644.0  0.103199     0.977839     0.122392     0.318059     0.603235   

      Majority Race  Majority Race Percentage Minority Race  \
0            White%                  0.804290        Black%   
2            White%                  0.691008        Black%   
3            White%                  0.879609        Black%   
4            White%                  0.796

In [2147]:
ruca_census_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 59873 entries, 0 to 71274
Data columns (total 30 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   State Code                     59873 non-null  object 
 1   County Code                    59873 non-null  object 
 2   Geoid Tract 2010               59873 non-null  object 
 3   Primary RUCA Code 2010         59873 non-null  int64  
 4   Community Type                 59873 non-null  object 
 5   Geoid Tract 2020               59873 non-null  object 
 6   Land Area (in Sq. Miles) 2020  59873 non-null  float64
 7   Census Tract Name              59873 non-null  object 
 8   Population                     59873 non-null  int64  
 9   Hous_Units                     59873 non-null  int64  
 10  Income                         59873 non-null  float64
 11  Poverty%                       59873 non-null  float64
 12  Employment%                    59873 non-null  floa

In [2151]:
# keeping only unique census tracts 
# ruca_census_unique.to_csv('ruca_census_final.csv')

The ruca_census_final.csv will be the csv file used for Logistic Regression model. We still need to feature engineer the target variable (bank desert status) using the census tracts data and bank locations data. 

### TIGER/line shapefiles
- This contains the census tract shapefiles for all 50 states in the US

In [8]:
tiger_shp_dir = r"C:\Users\jocel\OneDrive\Documents\University_of_Denver\COMP 4447 DS Tools 1\Final Project\Bank-Deserts-Phenomenon\data_collection\tiger_shp\tiger_shp_unzipped"
print(os.listdir(tiger_shp_dir))

['tl_2024_01_tract', 'tl_2024_02_tract', 'tl_2024_04_tract', 'tl_2024_05_tract', 'tl_2024_06_tract', 'tl_2024_08_tract', 'tl_2024_09_tract', 'tl_2024_10_tract', 'tl_2024_11_tract', 'tl_2024_12_tract', 'tl_2024_13_tract', 'tl_2024_15_tract', 'tl_2024_16_tract', 'tl_2024_17_tract', 'tl_2024_18_tract', 'tl_2024_19_tract', 'tl_2024_20_tract', 'tl_2024_21_tract', 'tl_2024_22_tract', 'tl_2024_23_tract', 'tl_2024_24_tract', 'tl_2024_25_tract', 'tl_2024_26_tract', 'tl_2024_27_tract', 'tl_2024_28_tract', 'tl_2024_29_tract', 'tl_2024_30_tract', 'tl_2024_31_tract', 'tl_2024_32_tract', 'tl_2024_33_tract', 'tl_2024_34_tract', 'tl_2024_35_tract', 'tl_2024_36_tract', 'tl_2024_37_tract', 'tl_2024_38_tract', 'tl_2024_39_tract', 'tl_2024_40_tract', 'tl_2024_41_tract', 'tl_2024_42_tract', 'tl_2024_44_tract', 'tl_2024_45_tract', 'tl_2024_46_tract', 'tl_2024_47_tract', 'tl_2024_48_tract', 'tl_2024_49_tract', 'tl_2024_50_tract', 'tl_2024_51_tract', 'tl_2024_53_tract', 'tl_2024_54_tract', 'tl_2024_55_tract',